src/lem/constrainer.rs

```diff
@@ -81,145 +116,390 @@ impl LEM {
         Ok(())
     }
 
-    fn on_concrete_path(concrete_path: &Option<Boolean>) -> Result<bool> {
-        match concrete_path {
-            None => Ok(true),
-            Some(concrete_path) => concrete_path
-                .get_value()
-                .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path")),
-        }
-    }
-
     fn allocate_and_inputize_input<'a, F: LurkField, CS: ConstraintSystem<F>>(
         &'a self,
         cs: &mut CS,
         store: &mut Store<F>,
-        witness: &Witness<F>,
+        valuation: &Valuation<F>,
         alloc_ptrs: &mut HashMap<&'a String, AllocatedPtr<F>>,
         idx: usize,
     ) -> Result<()> {
-        if alloc_ptrs.contains_key(&self.input[idx]) {
-            bail!("{} already allocated", &self.input[idx]);
-        }
         let alloc_ptr = Self::allocate_ptr(
-            &mut cs.namespace(|| format!("allocate input {}", &self.input[idx])),
-            &store.hydrate_ptr(&witness.input[idx])?,
+            cs,
+            &store.hydrate_ptr(&valuation.input[idx])?,
             &self.input[idx],
+            alloc_ptrs,
         )?;
         alloc_ptrs.insert(&self.input[idx], alloc_ptr.clone());
         Self::inputize_ptr(cs, &alloc_ptr, &self.input[idx])?;
         Ok(())
     }
 
-    /// Create R1CS constraints for LEM given an evaluation witness.
+    fn on_concrete_path(concrete_path: &Boolean) -> Result<bool> {
+        concrete_path
+            .get_value()
+            .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path"))
+    }
+
+    fn z_ptr_from_valuation<F: LurkField>(
+        concrete_path: &Boolean,
+        valuation: &Valuation<F>,
+        name: &String,
+        store: &mut Store<F>,
+    ) -> Result<ZPtr<F>> {
+        if Self::on_concrete_path(concrete_path)? {
+            let Some(ptr) = valuation.ptrs.get(name) else {
+                bail!("Couldn't retrieve {} from valuation", name);
+            };
+            store.hydrate_ptr(ptr)
+        } else {
+            Ok(ZPtr::dummy())
+        }
+    }
+
+    /// Accumulates slot data that will be used later to generate the constraints.
+    fn acc_slots_data_data<F: LurkField>(
+        concrete_path: Boolean,
+        slots_data: &mut Vec<SlotData<F>>,
+        img: AllocatedNum<F>,
+        preimg_vec: Vec<&AllocatedPtr<F>>,
```

Comment:
Preimages are always vectors or similar. This could just be `preimage` to keep reading this code simple.

Reply:
I see. This isn't a 'normal' preimage in terms of being a vector of field elements. However `_vec` completely misses the point of how it differs. The important point is that this contains (allocated) pointers.